### PR TITLE
Adding Flapper platform and handling servomotors

### PR DESCRIPTION
**Pull request by matejkarasek**
This PR adds a "flapper" platform, together with a "flapper deck", default settings and gain values.
To enable servomotors, which need a non-zero neutral value, additions were made also to motors.c and supervisor.c
(This is a clean version of #1123)

Finally, I include some cleanup after #1121.

### CODE REVIEW

File: bindings/setup.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
     "src/utils/src/num.c",
     "src/modules/src/controller_mellinger.c",
     "src/modules/src/power_distribution_quadrotor.c",
+    # "src/modules/src/power_distribution_flapper.c",
```
Comment (matejkarasek) [inline bindings/setup.py:31]: The Python bindings might require some extra work such that they function correctly with the Flapper...
If the second power distribution is included, the build test fails...

**Comment by tobbeanton** [comment]
Overall I think it looks good! There are two discussion points I can identify though:
1. The servo integration in motors.c is starting to make this file quite large. Maybe it is time to split it up and have a servo.c and maybe a higher level actuator.c file. This is quite a large change though...
2. There is a potential conflicting PR #1131 which is changing the external battery monitoring functionality a bit. Which one is easiest to merge first? Probably PR #1131 since it is smaller.

**Review by krichardsson — CHANGES_REQUESTED** [review]
I think this generally looks good! 
I mainly have some comments related to ifdefs based on platform, see comments in the code.

**Comment by matejkarasek** [comment]
> Overall I think it looks good! There are two discussion points I can identify though:

Thanks, @tobbeanton!

> 1. The servo integration in motors.c is starting to make this file quite large. Maybe it is time to split it up and have a servo.c and maybe a higher level actuator.c file. This is quite a large change though...

I agree. In fact I have started working on another PR that should enable more than 4 motors/servos:
https://github.com/flapper-drones/crazyflie-firmware/tree/more_motors

So how about if I introduce the servo & motor separation there?

> 2. There is a potential conflicting PR [Bigquad early access fix #1131](https://github.com/bitcraze/crazyflie-firmware/pull/1131) which is changing the external battery monitoring functionality a bit. Which one is easiest to merge first? Probably PR [Bigquad early access fix #1131](https://github.com/bitcraze/crazyflie-firmware/pull/1131) since it is smaller.

I agree it is easier to merge #1131 first

**Comment by matejkarasek** [comment]
> I think this generally looks good! I mainly have some comments related to ifdefs based on platform, see comments in the code.

Thanks for the quick review @krichardsson 
Will fix try to fix these soon

File: src/drivers/src/motors.c [hunk 2]
```diff
@@ -49,6 +50,8 @@ static bool motorSetEnable = false;
 static uint32_t motorPower[] = {0, 0, 0, 0};    // user-requested PWM signals
 static uint16_t motorPowerSet[] = {0, 0, 0, 0}; // user-requested PWM signals (overrides)
 static uint32_t motor_ratios[] = {0, 0, 0, 0};  // actual PWM signals
+uint32_t motor_ratios_stop[] = {0, 0, 0, 0}; // motor-stop PWM signals
```
[the remark this answers is not in the document]
Comment (matejkarasek) [reply src/drivers/src/motors.c:53]: yes, indeed...

File: src/modules/src/supervisor.c [hunk 3]
```diff
@@ -117,9 +118,11 @@ void supervisorUpdate(const sensorData_t *data)
   isFlying = isFlyingCheck();
 
   isTumbled = isTumbledCheck(data);
+  #ifndef SUPERVISOR_TUMBLE_CHECK_DISABLE
```
[the remark this answers is not in the document]
Comment (matejkarasek) [reply src/modules/src/supervisor.c:121]: True :)

Changed to `SUPERVISOR_TUMBLE_CHECK_ENABLE` which is set to `true` in platform_defaults.h and can be overridden in the defaults of the individual platforms

File: src/drivers/src/motors.c [hunk 4]
```diff
@@ -272,6 +275,21 @@ void motorsInit(const MotorPerifDef** motorMapSelect)
 
   isInit = true;
 
+// For platforms that use servos, set motor numbers and motor stop values
```
[the remark this answers is not in the document]
Comment (matejkarasek) [reply src/drivers/src/motors.c:278]: I suspected you may not like this :)

I now created two new functions declared in `power_distribution.h` and implemented in the c of the various power distributions. Would this work?

File: src/modules/interface/power_distribution.h [hunk 5]
```diff
@@ -33,4 +33,10 @@ void powerDistributionInit(void);
 bool powerDistributionTest(void);
 void powerDistribution(motors_thrust_t* motorPower, const control_t *control);
 
+#ifdef CONFIG_POWER_DISTRIBUTION_FLAPPER
```
[the remark this answers is not in the document]
Comment (matejkarasek) [reply src/modules/interface/power_distribution.h:36]: These were used in motors.c, so it was necessary. Not needed anymore...

**Comment by matejkarasek** [comment]
All the issues are fixed, pls let me know if this works.
And also whether I should do anything about #1131.
Thanks!

**Comment by matejkarasek** [comment]
@knmcguire Thanks for testing it out! 

Have you trimmed the servos (especially the pitch servo) after the firmware update? It is stored as a permanent parameter, and I have changed the name of the parameter group from "_flapper" to "flapper". So you have to do it after this firmware update...

**Comment by knmcguire** [comment]
Yes that's it! the flapper.servPitchNeutr had to be adjusted.  It was set to 50 but I had to change it to 58. Let me discuss with @krichardsson if there was anything else but if that is it, we'll merge it soon!

**Comment by krichardsson** [comment]
Looking nice and clean!
Merged
Welcome flapper!

**Comment by matejkarasek** [comment]
Wohoo! Awesome, thanks!


> Yes that's it! the flapper.servPitchNeutr had to be adjusted. It was set to 50 but I had to change it to 58. Let me discuss with @krichardsson if there was anything else but if that is it, we'll merge it soon!
The default value is close, but not exact, this varies from Flapper to Flapper (or from servo to servo, to be precise). The same applies for the yaw servo, though there it is less critical...

**Pull request merged**